From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Sat, 10 Oct 2026 21:42:08 +0200
Subject: [PATCH] Keep the time in single precision

The clocks (realtime, host_frametime, cl.time, sv.time and the network
timers) were doubles, which made every physics, view and particle
calculation with a time work in double precision, several times slower
in software than single precision. QuakeC already has the time as a
float. The clock starts with the DS, a float still resolves a
millisecond after more than four hours and a few after a day.
---
 WinQuake/client.h   |  8 ++++----
 WinQuake/common.c   |  2 +-
 WinQuake/console.c  |  2 +-
 WinQuake/host.c     | 24 ++++++++++++------------
 WinQuake/net.h      | 14 +++++++-------
 WinQuake/net_main.c | 16 ++++++++--------
 WinQuake/net_vcr.c  |  2 +-
 WinQuake/quakedef.h |  4 ++--
 WinQuake/r_nds.c    |  2 +-
 WinQuake/render.h   |  2 +-
 WinQuake/sbar_nds.c |  2 +-
 WinQuake/server.h   |  8 ++++----
 WinQuake/snd_nds.c  |  4 ++--
 WinQuake/sv_phys.c  |  2 +-
 WinQuake/sys.h      |  2 +-
 WinQuake/sys_nds.c  | 10 ++++++----
 WinQuake/vid_nds.c  |  2 +-
 17 files changed, 54 insertions(+), 52 deletions(-)

diff --git a/WinQuake/client.h b/WinQuake/client.h
index 7c59ee9..959dfd8 100644
--- a/WinQuake/client.h
+++ b/WinQuake/client.h
@@ -181,7 +181,7 @@ typedef struct
 	float		pitchvel;
 	qboolean	nodrift;
 	float		driftmove;
-	double		laststop;
+	float		laststop;
 
 	float		viewheight;
 	float		crouch;			// local amount for smoothing stepups
@@ -193,11 +193,11 @@ typedef struct
 	int			intermission;	// don't change view angle, full screen, etc
 	int			completed_time;	// latched at intermission start
 	
-	double		mtime[2];		// the timestamp of last two messages	
-	double		time;			// clients view of time, should be between
+	float		mtime[2];		// the timestamp of last two messages	
+	float		time;			// clients view of time, should be between
 								// servertime and oldservertime to generate
 								// a lerp point for other data
-	double		oldtime;		// previous cl.time, time-oldtime is used
+	float		oldtime;		// previous cl.time, time-oldtime is used
 								// to decay light values and smooth step ups
 	
 
diff --git a/WinQuake/common.c b/WinQuake/common.c
index c9925a6..544bb0a 100644
--- a/WinQuake/common.c
+++ b/WinQuake/common.c
@@ -350,7 +350,7 @@ int Q_atoi (char *str)
 
 float Q_atof (char *str)
 {
-	double			val;
+	float			val;
 	int             sign;
 	int             c;
 	int             decimal, total;
diff --git a/WinQuake/console.c b/WinQuake/console.c
index fef7c20..e9b326e 100644
--- a/WinQuake/console.c
+++ b/WinQuake/console.c
@@ -639,7 +639,7 @@ Con_NotifyBox
 */
 void Con_NotifyBox (char *text)
 {
-	double		t1, t2;
+	float		t1, t2;
 
 // during startup for sound / cd warnings
 	Con_Printf("\n\n\35\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\36\37\n");
diff --git a/WinQuake/host.c b/WinQuake/host.c
index fbf055f..1626a38 100644
--- a/WinQuake/host.c
+++ b/WinQuake/host.c
@@ -37,10 +37,10 @@ quakeparms_t host_parms;
 
 qboolean	host_initialized;		// true if into command execution
 
-double		host_frametime;
-double		host_time;
-double		realtime;				// without any filtering or bounding
-double		oldrealtime;			// last frame run
+float		host_frametime;
+float		host_time;
+float		realtime;				// without any filtering or bounding
+float		oldrealtime;			// last frame run
 int			host_framecount;
 
 int			host_hunklevel;
@@ -407,7 +407,7 @@ void Host_ShutdownServer(qboolean crash)
 	int		count;
 	sizebuf_t	buf;
 	char		message[4];
-	double	start;
+	float	start;
 
 	if (!sv.active)
 		return;
@@ -631,9 +631,9 @@ Runs all active servers
 */
 void _Host_Frame (float time)
 {
-	static double		time1 = 0;
-	static double		time2 = 0;
-	static double		time3 = 0;
+	static float		time1 = 0;
+	static float		time2 = 0;
+	static float		time3 = 0;
 	int			pass1, pass2, pass3;
 
 	if (setjmp (host_abortserver) )
@@ -674,8 +674,8 @@ void _Host_Frame (float time)
 	{
 		// the server runs at a lower rate, the client interpolates the
 		// entities in between
-		static double	server_time;
-		double			frametime = host_frametime;
+		static float	server_time;
+		float			frametime = host_frametime;
 
 		server_time += host_frametime;
 		if (server_time >= 1.0 / SERVER_FPS)
@@ -742,8 +742,8 @@ void _Host_Frame (float time)
 
 void Host_Frame (float time)
 {
-	double	time1, time2;
-	static double	timetotal;
+	float	time1, time2;
+	static float	timetotal;
 	static int		timecount;
 	int		i, c, m;
 
diff --git a/WinQuake/net.h b/WinQuake/net.h
index af46125..9fdbe79 100644
--- a/WinQuake/net.h
+++ b/WinQuake/net.h
@@ -119,9 +119,9 @@ struct qsockaddr
 typedef struct qsocket_s
 {
 	struct qsocket_s	*next;
-	double			connecttime;
-	double			lastMessageTime;
-	double			lastSendTime;
+	float			connecttime;
+	float			lastMessageTime;
+	float			lastSendTime;
 
 	qboolean		disconnected;
 	qboolean		canSend;
@@ -218,7 +218,7 @@ extern int		unreliableMessagesReceived;
 
 qsocket_t *NET_NewQSocket (void);
 void NET_FreeQSocket(qsocket_t *);
-double SetNetTime(void);
+float SetNetTime(void);
 
 
 #define HOSTCACHESIZE	8
@@ -263,7 +263,7 @@ qboolean IsID(struct qsockaddr *addr);
 //
 //============================================================================
 
-extern	double		net_time;
+extern	float		net_time;
 extern	sizebuf_t	net_message;
 extern	int			net_activeconnections;
 
@@ -313,12 +313,12 @@ void NET_Poll(void);
 typedef struct _PollProcedure
 {
 	struct _PollProcedure	*next;
-	double					nextTime;
+	float					nextTime;
 	void					(*procedure)();
 	void					*arg;
 } PollProcedure;
 
-void SchedulePollProcedure(PollProcedure *pp, double timeOffset);
+void SchedulePollProcedure(PollProcedure *pp, float timeOffset);
 
 extern	qboolean	serialAvailable;
 extern	qboolean	ipxAvailable;
diff --git a/WinQuake/net_main.c b/WinQuake/net_main.c
index 5f5d2cf..a0fad37 100644
--- a/WinQuake/net_main.c
+++ b/WinQuake/net_main.c
@@ -46,7 +46,7 @@ static qboolean	listening = false;
 qboolean	slistInProgress = false;
 qboolean	slistSilent = false;
 qboolean	slistLocal = true;
-static double	slistStartTime;
+static float	slistStartTime;
 static int		slistLastShown;
 
 static void Slist_Send(void);
@@ -90,9 +90,9 @@ qboolean recording = false;
 int	net_driverlevel;
 
 
-double			net_time;
+float			net_time;
 
-double SetNetTime(void)
+float SetNetTime(void)
 {
 	net_time = Sys_FloatTime();
 	return net_time;
@@ -449,7 +449,7 @@ NET_CheckNewConnections
 
 struct
 {
-	double	time;
+	float	time;
 	int		op;
 	long	session;
 } vcrConnect;
@@ -528,7 +528,7 @@ returns -1 if connection is invalid
 
 struct
 {
-	double	time;
+	float	time;
 	int		op;
 	long	session;
 	int		ret;
@@ -616,7 +616,7 @@ returns -1 if the connection died
 */
 struct
 {
-	double	time;
+	float	time;
 	int		op;
 	long	session;
 	int		r;
@@ -721,7 +721,7 @@ qboolean NET_CanSendMessage (qsocket_t *sock)
 
 int NET_SendToAll(sizebuf_t *data, int blocktime)
 {
-	double		start;
+	float		start;
 	int			i;
 	int			count = 0;
 	qboolean	state1 [MAX_SCOREBOARD];
@@ -955,7 +955,7 @@ void NET_Poll(void)
 }
 
 
-void SchedulePollProcedure(PollProcedure *proc, double timeOffset)
+void SchedulePollProcedure(PollProcedure *proc, float timeOffset)
 {
 	PollProcedure *pp, *prev;
 
diff --git a/WinQuake/net_vcr.c b/WinQuake/net_vcr.c
index ba8f40d..af51198 100644
--- a/WinQuake/net_vcr.c
+++ b/WinQuake/net_vcr.c
@@ -31,7 +31,7 @@ extern int vcrFile;
 
 static struct
 {
-	double	time;
+	float	time;
 	int		op;
 	long	session;
 }	next;
diff --git a/WinQuake/quakedef.h b/WinQuake/quakedef.h
index a72b2dd..912c555 100644
--- a/WinQuake/quakedef.h
+++ b/WinQuake/quakedef.h
@@ -303,11 +303,11 @@ extern	cvar_t		sys_nostdout;
 extern	cvar_t		developer;
 
 extern	qboolean	host_initialized;		// true if into command execution
-extern	double		host_frametime;
+extern	float		host_frametime;
 extern	byte		*host_basepal;
 extern	byte		*host_colormap;
 extern	int			host_framecount;	// incremented every frame, never reset
-extern	double		realtime;			// not bounded in any way, changed at
+extern	float		realtime;			// not bounded in any way, changed at
 										// start of every frame, never reset
 
 void Host_ClearMemory (void);
diff --git a/WinQuake/r_nds.c b/WinQuake/r_nds.c
index 783a744..905379c 100644
--- a/WinQuake/r_nds.c
+++ b/WinQuake/r_nds.c
@@ -2668,7 +2668,7 @@ static void R_DimView (void)
 
 void R_RenderView (void)
 {
-	double	time1 = 0;
+	float	time1 = 0;
 
 	if (!r_worldentity.model || !cl.worldmodel)
 		Sys_Error ("R_RenderView: NULL worldmodel");
diff --git a/WinQuake/render.h b/WinQuake/render.h
index b5f8382..c29cd4c 100644
--- a/WinQuake/render.h
+++ b/WinQuake/render.h
@@ -44,7 +44,7 @@ typedef struct entity_s
 
 	entity_state_t			baseline;		// to fill in defaults in updates
 
-	double					msgtime;		// time of last update
+	float					msgtime;		// time of last update
 	vec3_t					msg_origins[2];	// last two updates (0 is newest)	
 	vec3_t					origin;
 	vec3_t					msg_angles[2];	// last two updates (0 is newest)
diff --git a/WinQuake/sbar_nds.c b/WinQuake/sbar_nds.c
index 1bf43ff..a2c023a 100644
--- a/WinQuake/sbar_nds.c
+++ b/WinQuake/sbar_nds.c
@@ -591,7 +591,7 @@ THE PANELS
 // The time in the level, it stops when you die
 static int Sbar_LevelTime (void)
 {
-	static double	deathtime = -1;
+	static float	deathtime = -1;
 
 	if (cl.stats[STAT_HEALTH] > 0 || cl.time < deathtime)
 		deathtime = -1;
diff --git a/WinQuake/server.h b/WinQuake/server.h
index 2d4dfeb..cb97fc4 100644
--- a/WinQuake/server.h
+++ b/WinQuake/server.h
@@ -39,10 +39,10 @@ typedef struct
 	qboolean	paused;
 	qboolean	loadgame;			// handle connections specially
 
-	double		time;
+	float		time;
 	
 	int			lastcheck;			// used by PF_checkclient
-	double		lastchecktime;
+	float		lastchecktime;
 	
 	char		name[64];			// map name
 #ifdef QUAKE2
@@ -83,7 +83,7 @@ typedef struct client_s
 	qboolean		privileged;			// can execute any host command
 	qboolean		sendsignon;			// only valid before spawned
 
-	double			last_message;		// reliable messages must be sent
+	float			last_message;		// reliable messages must be sent
 										// periodically
 
 	struct qsocket_s *netconnection;	// communications handle
@@ -209,7 +209,7 @@ extern	client_t	*host_client;
 
 extern	jmp_buf 	host_abortserver;
 
-extern	double		host_time;
+extern	float		host_time;
 
 extern	edict_t		*sv_player;
 
diff --git a/WinQuake/snd_nds.c b/WinQuake/snd_nds.c
index 5a269c9..044f46b 100644
--- a/WinQuake/snd_nds.c
+++ b/WinQuake/snd_nds.c
@@ -170,11 +170,11 @@ static void HW_Start (int slot, int channel, sfxcache_t *sc)
 
 static sfxcache_t *SND_LoadSound (sfx_t *sfx)
 {
-	static double	retry;
+	static float	retry;
 	static struct
 	{
 		sfx_t	*sfx;
-		double	loaded, blocked;
+		float	loaded, blocked;
 	} loads[SND_LOADS];
 	sfxcache_t		*sc = Cache_Check (&sfx->cache);
 	int				i = ((uintptr_t) sfx >> 4) % SND_LOADS;
diff --git a/WinQuake/sv_phys.c b/WinQuake/sv_phys.c
index e13c0e4..a05afca 100644
--- a/WinQuake/sv_phys.c
+++ b/WinQuake/sv_phys.c
@@ -1571,7 +1571,7 @@ trace_t SV_Trace_Toss (edict_t *ent, edict_t *ignore)
 	trace_t	trace;
 	vec3_t	move;
 	vec3_t	end;
-	double	save_frametime;
+	float	save_frametime;
 //	extern particle_t	*active_particles, *free_particles;
 //	particle_t	*p;
 
diff --git a/WinQuake/sys.h b/WinQuake/sys.h
index 1b19721..a8e5abe 100644
--- a/WinQuake/sys.h
+++ b/WinQuake/sys.h
@@ -57,7 +57,7 @@ void Sys_EndTextConsole (void);
 
 void Sys_Quit (void);
 
-double Sys_FloatTime (void);
+float Sys_FloatTime (void);
 
 char *Sys_ConsoleInput (void);
 
diff --git a/WinQuake/sys_nds.c b/WinQuake/sys_nds.c
index 6de9d49..2d41103 100644
--- a/WinQuake/sys_nds.c
+++ b/WinQuake/sys_nds.c
@@ -266,10 +266,12 @@ void Sys_Quit (void)
 	exit (0);
 }
 
-double Sys_FloatTime (void)
+float Sys_FloatTime (void)
 {
-	// in double precision, constants are single precision (Makefile.nds)
-	return (double) tickGetCount () / TICK_FREQ;
+	// Seconds since the start in single precision, which the DS works out
+	// much faster than double precision and still resolves a millisecond
+	// after hours
+	return (float) tickGetCount () / TICK_FREQ;
 }
 
 char *Sys_ConsoleInput (void)
@@ -354,7 +356,7 @@ static void Sys_SetPlayerName (void)
 
 static int Sys_QuakeMain (void *arg)
 {
-	double	time, oldtime, newtime;
+	float	time, oldtime, newtime;
 
 	Host_Init (arg);
 	Sys_SetPlayerName ();
diff --git a/WinQuake/vid_nds.c b/WinQuake/vid_nds.c
index 2985b81..bd39406 100644
--- a/WinQuake/vid_nds.c
+++ b/WinQuake/vid_nds.c
@@ -572,7 +572,7 @@ void VID_Flush3D (void)
 // Frames per second of the last second, in the top right corner
 static void VID_DrawFPS (void)
 {
-	static double	lasttime;
+	static float	lasttime;
 	static int		frames, totalframes;
 	static char		text[16];
 	char			log[48];
